File: razer-sdk/src/test/java/net/booone/razersdk/animation/LayeredFrameTest.java
package net.booone.razersdk.animation;

import net.booone.razersdk.color.Color;
import net.booone.razersdk.color.StaticColor;
import net.booone.razersdk.sdk.RzKey;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;

import static org.junit.jupiter.api.Assertions.*;

class LayeredFrameTest {

    @Test
    void testEmptyFrameCreation() {
        //given
        final LayeredFrame layeredFrame = new LayeredFrame();

        //when
        final Frame actualFrame = layeredFrame.getFrame();

        //then
        assertTrue(actualFrame.getKeysToColors().isEmpty());
        assertFalse(layeredFrame.hasFrame());
        assertThrows(NoSuchElementException.class, layeredFrame::getFrame);
    }

    @Test
    void testStartWithFrame() {
        //when
        final LayeredFrame layeredFrame = new LayeredFrame();

        //then
        assertTrue(layeredFrame.hasFrame());
    }

    @Test
    void testLayerCreation() {
        //given
        final LayeredFrame layeredFrame = createTestLayer();
        final Map<RzKey, Color> actualColorMap = layeredFrame.getFrame().getKeysToColors();

        //then
        assertEquals(actualColorMap.get(RzKey.RZKEY_1), StaticColor.BLUE);
        assertEquals(actualColorMap.get(RzKey.RZKEY_3), StaticColor.BLUE);
        assertEquals(actualColorMap.get(RzKey.RZKEY_2), StaticColor.RED);
        assertEquals(3, actualColorMap.size());
        assertFalse(layeredFrame.hasFrame());
        assertThrows(NoSuchElementException.class, layeredFrame::getFrame);
    }

    private static LayeredFrame createTestLayer() {
        final SimpleFrame layerOne = new SimpleFrame(List.of(RzKey.RZKEY_1, RzKey.RZKEY_2, RzKey.RZKEY_3), StaticColor.BLUE);
        final SimpleFrame layerTwo = new SimpleFrame(RzKey.RZKEY_2, StaticColor.RED);
        final SimpleFrame transparentLayer = new SimpleFrame(List.of(RzKey.RZKEY_1, RzKey.RZKEY_2, RzKey.RZKEY_3, RzKey.RZKEY_4), StaticColor.NONE);

        //when
        final LayeredFrame layeredFrame = new LayeredFrame();
        layeredFrame.addFrame(layerOne);
        layeredFrame.addFrame(layerTwo);
        layeredFrame.addFrame(transparentLayer);
        return layeredFrame;
    }
}

File: test/razer-sdk/animation/LayeredFrame.test.ts
import {describe, expect, test} from 'vitest';
import {LayeredFrame} from '../../../src/razer-sdk/animation/LayeredFrame.js';
import {SimpleFrame} from '../../../src/razer-sdk/animation/SimpleFrame.js';
import {StaticColor} from '../../../src/razer-sdk/color/StaticColor.js';
import {RzKey} from '../../../src/razer-sdk/sdk/RzKey.js';

function createTestLayer(): LayeredFrame {
    const layerOne = new SimpleFrame([RzKey.RZKEY_1, RzKey.RZKEY_2, RzKey.RZKEY_3], StaticColor.BLUE);
    const layerTwo = new SimpleFrame(RzKey.RZKEY_2, StaticColor.RED);
    const transparentLayer = new SimpleFrame([RzKey.RZKEY_1, RzKey.RZKEY_2, RzKey.RZKEY_3, RzKey.RZKEY_4], StaticColor.NONE);

    //when
    const layeredFrame = new LayeredFrame();
    layeredFrame.addFrame(layerOne);
    layeredFrame.addFrame(layerTwo);
    layeredFrame.addFrame(transparentLayer);
    return layeredFrame;
}

describe('LayeredFrameTest', () => {
    test('testEmptyFrameCreation', () => {
        //given
        const layeredFrame = new LayeredFrame();

        //when
        const actualFrame = layeredFrame.pollFrame();

        //then
        expect(actualFrame.getKeysToColors().size).toBe(0);
        expect(layeredFrame.hasFrame()).toBe(false);
        expect(() => layeredFrame.pollFrame()).toThrow();
    });

    test('testStartWithFrame', () => {
        //when
        const layeredFrame = new LayeredFrame();

        //then
        expect(layeredFrame.hasFrame()).toBe(true);
    });

    test('testLayerCreation', () => {
        //given
        const layeredFrame = createTestLayer();
        const actualColorMap = layeredFrame.pollFrame().getKeysToColors();

        //then
        expect(actualColorMap.get(RzKey.RZKEY_1)).toBe(StaticColor.BLUE);
        expect(actualColorMap.get(RzKey.RZKEY_3)).toBe(StaticColor.BLUE);
        expect(actualColorMap.get(RzKey.RZKEY_2)).toBe(StaticColor.RED);
        expect(actualColorMap.size).toBe(3);
        expect(layeredFrame.hasFrame()).toBe(false);
        expect(() => layeredFrame.pollFrame()).toThrow();
    });
});
